app/code/Magento/Inventory/Observer/SourceItemsProcessor.php

```diff
@@ -0,0 +1,158 @@
+
+/**
+ * Class CarrierRequestDataHydrator
+ */
+class SourceItemsProcessor
+{
+    /**
+     * @var SearchCriteriaBuilderFactory
+     */
+    private $searchCriteriaBuilderFactory;
+
+    /**
+     * @var SourceItemRepositoryInterface
+     */
+    private $sourceItemRepository;
+
+    /**
+     * @var SourceItemInterfaceFactory
+     */
+    private $sourceItemFactory;
+
+    /**
+     * @var DataObjectHelper
+     */
+    private $dataObjectHelper;
+
+    /**
+     * @var SourceItemSaveInterface
+     */
+    private $sourceItemSave;
+
+    /**
+     * SourceItemsProcessor constructor
+     *
+     * @param SearchCriteriaBuilderFactory $searchCriteriaBuilderFactory
+     * @param SourceItemRepositoryInterface $sourceItemRepository
+     * @param SourceItemInterfaceFactory $sourceItemFactory
+     * @param DataObjectHelper $dataObjectHelper
+     * @param SourceItemSaveCommandInterface $sourceItemSave
+     */
+    public function __construct(
+        SearchCriteriaBuilderFactory $searchCriteriaBuilderFactory,
+        SourceItemRepositoryInterface $sourceItemRepository,
+        SourceItemInterfaceFactory $sourceItemFactory,
+        DataObjectHelper $dataObjectHelper,
+        SourceItemSaveInterface $sourceItemSave
+    ) {
+        $this->searchCriteriaBuilderFactory = $searchCriteriaBuilderFactory;
+        $this->sourceItemRepository = $sourceItemRepository;
+        $this->sourceItemFactory = $sourceItemFactory;
+        $this->dataObjectHelper = $dataObjectHelper;
+        $this->sourceItemSave = $sourceItemSave;
+    }
+
+    /**
+     * @param string $sku
+     * @param array $sourceItemsData
+     * @return void
+     * @throws InputException
+     */
+    public function process($sku, array $sourceItemsData)
+    {
+        $sourceItemsForDelete = $this->getCurrentSourceItemsMap($sku);
+        $sourceItemsForSave = [];
+
+        foreach ($sourceItemsData as $sourceItemData) {
+            $this->validateSourceItemData($sourceItemData);
+
+            $sourceId = $sourceItemData[SourceItemInterface::SOURCE_ID];
+            if (isset($sourceItemsForDelete[$sourceId])) {
+                $sourceItem = $sourceItemsForDelete[$sourceId];
+            } else {
+                /** @var SourceItemInterface $relation */
+                $sourceItem = $this->sourceItemFactory->create();
+            }
+
+            $sourceItemData[SourceItemInterface::SKU] = $sku;
+            $this->dataObjectHelper->populateWithArray($sourceItem, $sourceItemData, SourceItemInterface::class);
+
+            $sourceItemsForSave[] = $sourceItem;
+            unset($sourceItemsForDelete[$sourceId]);
+        }
+        $this->saveSourceItems($sourceItemsForSave);
+        $this->deleteSourceItems($sourceItemsForDelete);
+    }
+
+    /**
+     * Key is source id, value is Source Item
+     *
+     * @param string $sku
+     * @return SourceItemInterface[]
+     */
+    private function getCurrentSourceItemsMap($sku)
```

Comment:
You can see this in method description
"Key is source id, value is Source Item"